ls1intum/Artemis
# Added validation for the file extensions in a file upload

### Checklist

### Motivation and Context
Fixes #1019
Right now, invalid extensions can be specified in the "File Pattern" field of the File Upload exercises, which leads to an error.

### Description
This PR adds a validation which proves the specified extensions against a list of valid extensions and displays an error message when the input is invalid.
The input need to be at least 2 characters long. 
Whitespaces are allowed (and removed once the form is saved). 
A valid list expects the extensions to be separated by commas (not whitespaces).
The input is case-insensitive. 

### Steps for Testing


1. Log in to Artemis as an instructor
2. Navigate to Course Administration
3. Navigate to Exercises of a Course
4. Create/Edit a File Upload Exercise 
5. Specify an invalid list of extensions in the File Pattern field (e.g. pdf, doxfg)
- You should be able to see a red stripe on the left side of the File Pattern field and an error message below it; and you should not be able to save the exercise
6. Specify a valid list of extensions in the File Pattern field (e.g. pdf, docx)
- You should be able to see a green stripe on the left side of the File Pattern field and you should be able to save the exercise

### Screenshots

Invalid extensions

![invalid extension](https://user-images.githubusercontent.com/33368508/70391049-2c480c80-19d1-11ea-914d-9ab977452d25.png)

![whitespaces not allowed](https://user-images.githubusercontent.com/33368508/70870045-292fbc00-1f8f-11ea-8e9f-dd60e629522b.png)


Valid extensions

![valid extensions](https://user-images.githubusercontent.com/33368508/70391052-3407b100-19d1-11ea-9074-9c6218948715.png)

![whitespaces allowed](https://user-images.githubusercontent.com/33368508/70870049-2cc34300-1f8f-11ea-865d-bc836c947994.png)

### Additional information

**Typeahead**
The suggestion to remove the free input field and use a drop-down menu/ a typeahead is not a part of this PR and can/will be addressed in [issue 1156](https://github.com/ls1intum/Artemis/issues/1156).

**Custom Validator**
There are 3 ways to create a custom validator: using _pattern_, _formControl_ or introducing the validator as a _directive_. 
 - @krusche suggested that I do not use regex validation, but validate against a list of valid extensions -> this is why the first option is not used. 
 - It is not a good idea to specify a form in another form which would have been the case if I specified the formControl for the filePattern attribute only; and I did not want to make too many changes (and rewrite the form so that it includes all attributes) that can lead to other problems -> second option: out.
 - So that left me with the directive. 

**Centralizing filePatterns in application-{env}.yml**
- As @ungaralex suggested, I centralized the list of exensions in the yml-files, so that there is no strange behavior on the client and the server side. (Such behavior can manifest if the client and the server use different lists for the validation.) 
- For accessing the filePatterns property on the server side, there are 3 options: using _'@ Value', '@PropertySource'_ and _'@ConfigurationProperties'_. I tried using the first one, but I was not able to get the value and was getting the error "Injection of autowired dependencies failed; nested exception is java.lang.IllegalArgumentException: Could not resolve placeholder in value".  As far as I read, there are some situations when you are not able to get the value or it is set to null and it is better to use the third option.
